tensorflow/contrib/rnn/python/ops/rnn_cell.py

```diff
@@ -19,16 +19,90 @@
 from tensorflow.python.framework import ops
+from tensorflow.python.framework.load_library import load_op_library
 from tensorflow.python.ops import array_ops
 from tensorflow.python.ops import clip_ops
+from tensorflow.python.ops import init_ops
 from tensorflow.python.ops import math_ops
 from tensorflow.python.ops import nn_ops
 from tensorflow.python.ops import rnn_cell
 from tensorflow.python.ops import variable_scope as vs
 from tensorflow.python.ops.math_ops import sigmoid
 from tensorflow.python.ops.math_ops import tanh
+from tensorflow.python.platform import resource_loader
+
+
+LSTM_OPS_FILE = "_lstm_ops.so"
+
+_lstm_ops = None
+_ops_lock = threading.Lock()
+
+
+@ops.RegisterShape("LSTMCellBlock")
+def _LSTMCellBlockShape(op):
+  batch_size = op.inputs[0].get_shape().with_rank(2)[0].value
+  cell_size = op.get_attr("cell_size")
+
+  return (tensor_shape.TensorShape([batch_size, cell_size]),
+          tensor_shape.TensorShape([batch_size, cell_size]),
+          tensor_shape.TensorShape([batch_size, cell_size]),
+          tensor_shape.TensorShape([batch_size, cell_size]),
+          tensor_shape.TensorShape([batch_size, cell_size]),
+          tensor_shape.TensorShape([batch_size, cell_size]),
+          tensor_shape.TensorShape([batch_size, cell_size * 2]),
+          tensor_shape.TensorShape([batch_size, cell_size]))
+
+
+@ops.RegisterGradient("LSTMCellBlock")
+def _LSTMCellBlockGrad(op, *grad):
+  x = op.inputs[0]
+  states_prev = op.inputs[1]
+  w = op.inputs[2]
+  b = op.inputs[3]
+
+  i = op.outputs[0]
+  cs = op.outputs[1]
+  f = op.outputs[2]
+  o = op.outputs[3]
+  ci = op.outputs[4]
+  co = op.outputs[5]
+  # states = op.outputs[6]
+  h = op.outputs[7]
+
+  states_grad = grad[6]
+  h_grad = grad[7]
+
+  parallel_dw = op.get_attr("parallel_dw")
+
+  (x_grad, states_prev_grad, w_grad, b_grad, dicfo, xh) = gen_lstm_ops._lstm_cell_block_grad(
+      x, states_prev, w, b, i, cs, f, o, ci, co, h, states_grad, h_grad,
+      cell_size=op.get_attr("cell_size"), bprop_dx=op.get_attr("bprop_dx"),
+      parallel_dw=parallel_dw)
+
+  if parallel_dw:
+    w_grad = math_ops.matmul(xh, dicfo, transpose_a=True)
+    b_grad = nn_ops.bias_add_grad(dicfo)
+  return (x_grad, states_prev_grad, w_grad, b_grad)
+
+
+@ops.RegisterShape("LSTMCellBlockGrad")
+def _LSTMCellBlockGradShape(op):
+  batch_size = op.inputs[0].get_shape().with_rank(2)[0].value
+  input_size = op.inputs[0].get_shape().with_rank(2)[1].value
+  cell_size = op.get_attr("cell_size")
+
+  return [tensor_shape.TensorShape([batch_size, input_size]),
+          tensor_shape.TensorShape([batch_size, cell_size * 2]),
+          tensor_shape.TensorShape([input_size + cell_size, cell_size * 4]),
+          tensor_shape.TensorShape([cell_size * 4]),
+          tensor_shape.TensorShape([batch_size, cell_size * 4]),
+          tensor_shape.TensorShape([batch_size, input_size + cell_size])]
 
 
 def _get_concat_variable(name, shape, dtype, num_shards):
```

Comment:
that code isn't mine, it was from the existing rnn_cell.py which I presume
was for TimeFreqLSTMCell which I presume is from Tara?

--
William Chan
Carnegie Mellon University
(650) 450-9455
williamchan.ca

On Fri, May 6, 2016 at 3:07 PM, ebrevdo <notifications@github.com> wrote:

> In tensorflow/contrib/rnn/python/ops/rnn_cell.py
> <https://github.com/tensorflow/tensorflow/pull/2002#discussion_r62397359>:
>
> > +    b_grad = nn_ops.bias_add_grad(dicfo)
> > +  return (x_grad, states_prev_grad, w_grad, b_grad)
> > +
> > +
> > +@ops.RegisterShape("LSTMCellBlockGrad")
> > +def _LSTMCellBlockGradShape(op):
> > +  batch_size = op.inputs[0].get_shape().with_rank(2)[0].value
> > +  input_size = op.inputs[0].get_shape().with_rank(2)[1].value
> > +  cell_size = op.get_attr("cell_size")
> > +
> > +  return [tensor_shape.TensorShape([batch_size, input_size]),
> > +          tensor_shape.TensorShape([batch_size, cell_size * 2]),
> > +          tensor_shape.TensorShape([input_size + cell_size, cell_size * 4]),
> > +          tensor_shape.TensorShape([cell_size * 4]),
> > +          tensor_shape.TensorShape([batch_size, cell_size * 4]),
> > +          tensor_shape.TensorShape([batch_size, input_size + cell_size])]
> >
> >
> >  def _get_concat_variable(name, shape, dtype, num_shards):
>
> fyi see the new get_variable partitioner behavior. you no longer need to
> worry about sharding yourself.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/2002/files/727e0c76bed212fdf240a221e577e1b1a86848d5#r62397359>
>